File: litedb-python/cluster_client.py
"""
cluster_client.py — a client for the distributed database.

A client can contact ANY node: the node routes single-key ops to the owning shard's leader and
coordinates cross-shard transactions itself. So the client just needs to reach one live node; it
shuffles the node list and retries the next on failure (which is also how it rides through a node
being killed mid-demo).
"""

import random
from typing import Optional

from cluster_config import NODES
from rpc import RPCClient


class ClusterClient:
    def __init__(self, timeout: float = 4.0):
        self._rpc = RPCClient(timeout=timeout)
        self._nodes = list(NODES.keys())

    def _call(self, method: str, payload: dict) -> dict:
        order = self._nodes[:]
        random.shuffle(order)
        last = {"ok": False, "error": "all_nodes_unreachable"}
        for nid in order:
            host, port = NODES[nid]
            resp = self._rpc.call(host, port, method, payload)
            if resp.get("ok"):
                return resp["result"]
            last = {"ok": False, "error": resp.get("error", "rpc_failed")}
        return last

    # ---- KV API -----------------------------------------------------------

    def put(self, key: str, value: str) -> dict:
        return self._call("put", {"key": key, "value": value})

    def delete(self, key: str) -> dict:
        return self._call("put", {"key": key, "value": None})

    def get(self, key: str, read_ts: Optional[int] = None) -> Optional[str]:
        res = self._call("get", {"key": key, "read_ts": read_ts})
        return res.get("value") if res.get("ok") else None

    def get_full(self, key: str, read_ts: Optional[int] = None) -> dict:
        return self._call("get", {"key": key, "read_ts": read_ts})

    # ---- transactions -----------------------------------------------------

    def begin(self) -> Optional[int]:
        res = self._call("begin", {})
        return res.get("read_ts") if res.get("ok") else None

    def txn(self, writes: dict, read_ts: Optional[int] = None) -> dict:
        """Atomically apply {key: value_or_None}. Spans shards via 2PC when needed."""
        return self._call("txn", {"writes": writes, "read_ts": read_ts})

    # ---- introspection ----------------------------------------------------

    def status(self) -> list[dict]:
        out = []
        for nid in self._nodes:
            host, port = NODES[nid]
            resp = self._rpc.call(host, port, "status", {})
            out.append(resp["result"] if resp.get("ok") else {"node": nid, "alive": False})
        return out

File: litedb-java/src/main/java/com/litedb/cluster/ClusterClient.java
package com.litedb.cluster;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * ClusterClient — a client for the distributed database. It can contact ANY node: the node routes
 * single-key ops to the owning shard's leader and coordinates cross-shard transactions. So the
 * client just shuffles the node list and retries the next on failure (which also rides through a
 * node being killed mid-demo).
 */
public final class ClusterClient {
    private final Rpc.Client rpc = new Rpc.Client(4000);
    private final List<String> nodes = ClusterConfig.nodeIds();

    @SuppressWarnings("unchecked")
    private Map<String, Object> call(String method, Map<String, Object> payload) {
        List<String> order = new ArrayList<>(nodes);
        Collections.shuffle(order);
        Map<String, Object> last = NodeServer.resp("ok", false, "error", "all_nodes_unreachable");
        for (String nid : order) {
            Map<String, Object> r = rpc.call(ClusterConfig.HOST, ClusterConfig.port(nid), method, payload);
            if (Boolean.TRUE.equals(r.get("ok"))) return (Map<String, Object>) r.get("result");
            last = NodeServer.resp("ok", false, "error", r.get("error"));
        }
        return last;
    }

    public Map<String, Object> put(String key, String value) {
        return call("put", NodeServer.resp("key", key, "value", value));
    }

    public Map<String, Object> delete(String key) {
        return call("put", NodeServer.resp("key", key, "value", null));
    }

    public String get(String key) {
        Map<String, Object> r = call("get", NodeServer.resp("key", key, "read_ts", null));
        return Boolean.TRUE.equals(r.get("ok")) ? (String) r.get("value") : null;
    }

    public String get(String key, long readTs) {
        Map<String, Object> r = call("get", NodeServer.resp("key", key, "read_ts", readTs));
        return Boolean.TRUE.equals(r.get("ok")) ? (String) r.get("value") : null;
    }

    public Map<String, Object> getFull(String key) {
        return call("get", NodeServer.resp("key", key, "read_ts", null));
    }

    public Long begin() {
        Map<String, Object> r = call("begin", new LinkedHashMap<>());
        return Boolean.TRUE.equals(r.get("ok")) ? ((Number) r.get("read_ts")).longValue() : null;
    }

    public Map<String, Object> txn(Map<String, Object> writes, Long readTs) {
        return call("txn", NodeServer.resp("writes", writes, "read_ts", readTs));
    }

    public List<Map<String, Object>> status() {
        List<Map<String, Object>> out = new ArrayList<>();
        for (String nid : nodes) {
            Map<String, Object> r = rpc.call(ClusterConfig.HOST, ClusterConfig.port(nid), "status", new LinkedHashMap<>());
            if (Boolean.TRUE.equals(r.get("ok"))) {
                @SuppressWarnings("unchecked")
                Map<String, Object> res = (Map<String, Object>) r.get("result");
                out.add(res);
            } else {
                out.add(NodeServer.resp("node", nid, "alive", false));
            }
        }
        return out;
    }
}
